gateway/src/apicast/policy/3scale_batcher/keys_helper.lua

```diff
@@ -0,0 +1,78 @@
+local ipairs = ipairs
+local format = string.format
+local insert = table.insert
+local concat = table.concat
+local sort = table.sort
+local unpack = table.unpack
+local ngx_re = ngx.re
+
+local _M = {}
+
+local function creds_part_in_key(creds)
+  if creds.app_id and creds.app_key then
+    return format("app_id:%s,app_key:%s", creds.app_id, creds.app_key)
+  elseif creds.user_key then
+    return format("user_key:%s", creds.user_key)
+  elseif creds.access_token then
+    return format("access_token:%s", creds.access_token)
+  end
+end
+
+local function metrics_part_in_key(usage)
+  local usages = {}
```

Comment:
You can preallocate the table by using `table.new`.

It is exposed in `resty.core.base` as `new_tab`. The params are the same as `lua_createtable`

Reply:
Done.